RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,214 @@
+    auto next_state = RandBLAS::sparse::fill_sparse(S);
+
+    // step 1.2: sketch the data matrix
+    int64_t lda_sk;
+    if (layout_A == blas::Layout::RowMajor) {
+        assert(lda >= n);
+        lda_sk = n;
+    } else {
+        assert(lda >= m);
+        lda_sk = d;
+    }
+    RandBLAS::util::safe_scal(d*n, 0.0, buff_A_sk, 1);
+    RandBLAS::sparse::lskges<T>(
+        layout_A,
+        blas::Op::NoTrans,
+        blas::Op::NoTrans,
+        d, n, m,
+        1.0, S, 0, 0,
+        buff_A, lda,
+        0.0, buff_A_sk, lda_sk
+    );
+
+    // step 2: compute SVD of sketch
+    T* ignore = nullptr;
+    if (layout_A == blas::Layout::RowMajor) {
+        //
+        //      buff_A_sk is stored in row-major format; we want its singular values
+        //      and right singular vectors.
+        //
+        //      We get around LAPACK's column-major restriction by using the fact that
+        //          memory(A_sk, RowMajor) == memory(transpose(A_sk), ColumnMajor).
+        //      So we'll tell LAPACK that we want the singular values and left
+        //      singular vectors of tranpose(A_sk).
+        //
+        lapack::Job jobu = lapack::Job::OverwriteVec;
+        lapack::Job jobvt = lapack::Job::NoVec;
+        lapack::gesvd(jobu, jobvt, n, d, buff_A_sk, n, sigma_sk, ignore, 1, ignore, 1);
+        // interpret buff_A_sk in row-major
+    } else {
+        lapack::Job jobu = lapack::Job::NoVec;
+        lapack::Job jobvt = lapack::Job::SomeVec;
+        T *Vt = new T[n * n]{};
+        lapack::gesvd(jobu, jobvt, d, n, buff_A_sk, d, sigma_sk, ignore, 1, Vt, n);
+        for (int i = 0; i < n; ++i) {
+            // buff_A_sk is now just scratch space that points to V_sk.
+            // We just computed its SVD as a d-by-n matrix in column-major order.
+            // Now we want to write to the first n^2 elements of buff_A_sk with
+            // n "n-vectors" in column-major order.
+            blas::copy(n, &Vt[i], n,  &buff_A_sk[i*n], 1);
+        }
+        delete[] Vt;
+    }
+    return next_state;
+}
+
+
+/**
+ * Accepts the right singular vectors and singular values of some tall
+ * matrix "T", along with a regularization parameter mu.
+ * 
+ * This function overwrites the the provided right singular vectors with
+ * a matrix "M" so that if
+ * 
+ *          T_aug := [T; sqrt(mu)*I]
+ * 
+ * then T_aug * M is column-orthonormal. Such a matrix M is called an
+ * _orthogonalizer_ of T_aug.
+ * 
+ * A thresholding scheme is applied to infer numerical rank of T_aug.
+ * 
+ * @param[in] n
+ *      The number of rows and columns in V.
+ * 
+ * @param[in,out] V
+ *      A buffer of size >= n*n, read in column-major order.
+ * 
+ *      On entry, the columns of V are the right singular vectors of some
+ *      tall matrix T.
+ * 
+ *      On exit, the i-th column of V is scaled down by
+ *              1/sqrt(sigma[i]^2 + mu)
+ *      up until i = rank-1, where rank is the return value of this function.
```

Comment:
Indexing here is a little confusing... I assume this means i = 0, ... , rank - 1 but i do worry that the mix of 0- and 1- indexing in math libraries means we may want to just do a small phrasing change to 
```
 *      On exit, the i-th column of V is scaled down by
 *              1/sqrt(sigma[i]^2 + mu)
 *      for i = 0, 1, ..., rank-1, where rank is the return value of this function.
```

Reply:
Excellent observation. I'll make this change.